### PR TITLE
✨  support NodeRegistrationOptions.IgnorePreflightErrors

### LINKED ISSUE #4581: Add IgnorePreflightErrors to KubeadmConfig types
**What steps did you take and what happened:**

Based on Slack thread: https://kubernetes.slack.com/archives/C8TSNPY4T/p1619121490147500

We saw that there were some fields missing which are in upstream kubeadm v1beta2. At a first glance we're missing:
* CertificateKey
* IgnorePreflightErrors

The theory in Slack was that these fields were added after we forked the types

**What did you expect to happen:**
I would have expected to be able to confgure (at least) `IgnorePreflightErrors`. 

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

**Environment:**

- Cluster-api version:
- Minikube/KIND version:
- Kubernetes version: (use `kubectl version`):
- OS (e.g. from `/etc/os-release`):

/kind bug
[One or more /area label. See https://github.com/kubernetes-sigs/cluster-api/labels?q=area for the list of labels]

### PR DESCRIPTION
**Pull request by sbueringer**
Signed-off-by: Stefan Büringer buringerst@vmware.com

**What this PR does / why we need it**:
For details see linked issue

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #4581

### CODE REVIEW

**Comment by sbueringer** [comment]
/cc @vincepri @fabriziopandini 
If I'm not mistaken, the IgnorePreflightErrors field was already added with kubeadm v1beta2 (I only had to add it to our structs not to our copies of the v1beta2 structs). 

I assume as we stopped supporting kubeadm v1beta1 with v1alpha4 (i.e. Kubernetes versions < 1.15), it doesn't make sense to add validation for the v1beta1 case?

File: controlplane/kubeadm/api/v1alpha3/conversion_test.go [hunk 1]
```diff
@@ -78,3 +80,11 @@ func kubeadmClusterConfigurationFuzzer(obj *kubeadmv1beta1.ClusterConfiguration,
 	// ClusterConfiguration.UseHyperKubeImage has been removed in v1alpha4, so setting it to false in order to avoid v1alpha3 --> v1alpha4 --> v1alpha3 round trip errors.
 	obj.UseHyperKubeImage = false
 }
+
+func kubeadmNodeRegistrationOptionsFuzzer(obj *kubeadmbootstrapv1alpha4.NodeRegistrationOptions, c fuzz.Continue) {
+	c.FuzzNoCustom(obj)
+
+	// NodeRegistrationOptions.IgnorePreflightErrors does not exist in kubeadm v1beta1 API, so setting it to nil in order to avoid
+	// v1alpha4 --> v1beta1 -> v1alpha4 round trip errors.
```
Comment (fabriziopandini) [inline controlplane/kubeadm/api/v1alpha3/conversion_test.go:88]: nit: should this be
```suggestion
	// v1alpha4 --> v1alpha3 -> v1alpha4 round trip errors.
```

File: bootstrap/kubeadm/api/v1alpha3/conversion_test.go [hunk 2]
```diff
@@ -68,3 +70,11 @@ func clusterConfigurationFuzzer(obj *v1beta1.ClusterConfiguration, c fuzz.Contin
 	// ClusterConfiguration.UseHyperKubeImage has been removed in v1alpha4, so setting it to false in order to avoid v1beta1 --> v1alpha4 --> v1beta1 round trip errors.
 	obj.UseHyperKubeImage = false
 }
+
+func kubeadmNodeRegistrationOptionsFuzzer(obj *kubeadmbootstrapv1alpha4.NodeRegistrationOptions, c fuzz.Continue) {
+	c.FuzzNoCustom(obj)
+
+	// NodeRegistrationOptions.IgnorePreflightErrors does not exist in kubeadm v1beta1 API, so setting it to nil in order to avoid
+	// v1alpha4 --> v1beta1 -> v1alpha4 round trip errors.
```
Comment (fabriziopandini) [inline bootstrap/kubeadm/api/v1alpha3/conversion_test.go:78]: nit: should this be 
```suggestion
	// v1alpha4 --> v1alpha3 -> v1alpha4 round trip errors.
```

**Review by fabriziopandini — COMMENTED** [review]
> If I'm not mistaken, the IgnorePreflightErrors field was already added with kubeadm v1beta2

You are right; given that it already existed in v1beta2 and now also in v1alpha4, what about cleaning up following custom conversion func https://github.com/kubernetes-sigs/cluster-api/blob/db4736853644a6440a1b2eea990d7af99a29d992/bootstrap/kubeadm/types/v1beta2/conversion.go#L73-L96
Eventually also v1beta2 <-> v1alpha4 conversion tests can now be cleaned up (https://github.com/kubernetes-sigs/cluster-api/blob/29333cd0802cf7bec32b57455c19262ab3d7b5b0/bootstrap/kubeadm/types/v1beta2/conversion_test.go#L71-L76)

**Comment by sbueringer** [comment]
@fabriziopandini dropped the conversion and fuzzer func for IgnorePreflight, ptal

**Comment by sbueringer** [comment]
@fabriziopandini *ping* In case you have some time. Not sure who else wants to review / to cc.

File: bootstrap/kubeadm/types/v1beta1/conversion_test.go [hunk 3]
```diff
@@ -77,3 +79,11 @@ func clusterConfigurationFuzzer(obj *ClusterConfiguration, c fuzz.Continue) {
...
 	obj.UseHyperKubeImage = false
 }
+
+func kubeadmNodeRegistrationOptionsFuzzer(obj *kubeadmbootstrapv1alpha4.NodeRegistrationOptions, c fuzz.Continue) {
+	c.FuzzNoCustom(obj)
+
+	// NodeRegistrationOptions.IgnorePreflightErrors does not exist in kubeadm v1beta1 API, so setting it to nil in order to avoid
+	// v1alpha4 --> v1beta1 -> v1alpha4 round trip errors.
+	obj.IgnorePreflightErrors = nil
+}
```
Comment (vincepri) [inline bootstrap/kubeadm/types/v1beta1/conversion_test.go:89]: Shouldn't we restore the field during conversion instead of dropping it?

Comment (sbueringer) [reply bootstrap/kubeadm/types/v1beta1/conversion_test.go:89]: @vincepri 
Very nice finding. I implemented it for the top-level types: KubeadmConfig, KubeadmConfigTemplate and KubeadmControlPlane.

Given that in the current case we convert between InitConfigurations and JoinConfigurations and they don't have ObjectMeta I'm not sure if it's possible in this case.

**Comment by sbueringer** [comment]
@vincepri PTAL :) (push --force because of rebase onto current main)

**Comment by sbueringer** [comment]
/test pull-cluster-api-test-main 
flakes

**Review by vincepri — APPROVED** [review]
/approve

**Comment by sbueringer** [comment]
@vincepri @fabriziopandini squashed the commits, so I need a lgtm again

**Pull request merged**